src/vm_pages.rs

```diff
@@ -1390,10 +1390,10 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
         let converted = self
             .inner
             .root
-            .get_invalidated_pages(page_addr, num_pages * page_size as u64, |addr| {
+            .get_invalidated_pages(page_addr, num_pages * page_size as u64, |addr, ps| {
                 self.inner.page_tracker.is_converted_page(
                     addr,
-                    PageSize::Size4k,
+                    ps,
                     self.inner.page_owner_id,
                     P::mem_type(),
                     version,
```

Comment:
for any of these functions where we're trying to create a hugepage out of 4k pages, we need to check that the pages that are mapped are contiguous and suitably aligned. i.e. the base GPA and SPA that it maps are aligned to the hugepage size, and that for each SPA we encounter in the page table it is contiguous to the one previous.

Reply:
Ah yes we do need to check it's contiguous and suitably aligned. I think this check should be handled directly inside `get_invalidated_pages()`, very similar to what I did with `promote_page()`. Or do you have another approach in mind?

Reply:
If I do it in the closure, that means I have to update the closure signature so that it would provide the previous address along with the current one. That means the logic of checking pages are contiguous and aligned must be implemented in multiple places. I was thinking about passing an extra parameter to the function `get_invalidated_pages()` and `get_mapped_pages()` so that they would know if the extra check for contiguous pages must be performed.

Reply:
> If I do it in the closure, that means I have to update the closure signature so that it would provide the previous address along with the current one.

can't the closure mutably capture a previous phys addr?

Reply:
Let me have a try!